File: packages/ts/src/query/jsonrpc.ts
/**
 * Generic JSON-RPC 2.0 client.
 *
 * Manages request ids (auto-incrementing per client), serializes the standard
 * `{ jsonrpc, id, method, params }` envelope, and maps JSON-RPC error responses
 * to a typed {@link JsonRpcError}. It is the transport base for both the EVM
 * `eth_*`/`net_*`/`web3_*` methods and the QoreChain `qor_*` namespace.
 *
 * The underlying `fetch` is injectable via {@link HttpOptions} so callers and
 * tests can supply their own transport.
 */

import { postJsonRpc, type HttpOptions } from "./http";

/** A JSON-RPC 2.0 error object. */
export interface JsonRpcErrorObject {
  code: number;
  message: string;
  data?: unknown;
}

/** A JSON-RPC 2.0 response envelope. */
export interface JsonRpcResponse<T> {
  jsonrpc: "2.0";
  id: number | string | null;
  result?: T;
  error?: JsonRpcErrorObject;
}

/** Thrown when a JSON-RPC response carries an `error` member. */
export class JsonRpcError extends Error {
  /** JSON-RPC error code. */
  readonly code: number;
  /** Optional implementation-defined error data. */
  readonly data?: unknown;

  constructor(code: number, message: string, data?: unknown) {
    super(message);
    this.name = "JsonRpcError";
    this.code = code;
    this.data = data;
    Object.setPrototypeOf(this, JsonRpcError.prototype);
  }
}

/** Options for {@link JsonRpcClient}. */
export type JsonRpcClientOptions = HttpOptions;

/** A minimal JSON-RPC 2.0 client over HTTP POST. */
export class JsonRpcClient {
  protected readonly url: string;
  protected readonly opts: JsonRpcClientOptions;
  private nextId = 1;

  constructor(url: string, opts: JsonRpcClientOptions = {}) {
    this.url = url;
    this.opts = opts;
  }

  /**
   * Invoke a JSON-RPC method and return its `result`.
   *
   * @param method - Wire method name (e.g. `"eth_chainId"`).
   * @param params - Positional params; defaults to an empty array.
   * @throws {@link JsonRpcError} when the response contains an `error`.
   * @throws {@link QoreHttpError} on a non-2xx transport response.
   */
  async call<T>(method: string, params: unknown[] = []): Promise<T> {
    const id = this.nextId++;
    const body = { jsonrpc: "2.0", id, method, params };
    const res = await postJsonRpc<JsonRpcResponse<T>>(this.url, body, this.opts);
    if (res.error) {
      throw new JsonRpcError(res.error.code, res.error.message, res.error.data);
    }
    return res.result as T;
  }

  /** `eth_chainId` — the chain id as a hex quantity string. */
  ethChainId(): Promise<string> {
    return this.call<string>("eth_chainId", []);
  }

  /** `eth_blockNumber` — the latest block height as a hex quantity string. */
  ethBlockNumber(): Promise<string> {
    return this.call<string>("eth_blockNumber", []);
  }

  /** `eth_getBalance` — wei balance of `address` as a hex quantity string. */
  ethGetBalance(address: string, block: string = "latest"): Promise<string> {
    return this.call<string>("eth_getBalance", [address, block]);
  }

  /** `net_version` — the network id as a decimal string. */
  netVersion(): Promise<string> {
    return this.call<string>("net_version", []);
  }

  /** `web3_clientVersion` — the node client version string. */
  web3ClientVersion(): Promise<string> {
    return this.call<string>("web3_clientVersion", []);
  }
}

File: packages/rust/src/query/jsonrpc.rs
//! Minimal JSON-RPC 2.0 client over HTTP POST.

use crate::error::{Error, Result};
use crate::query::DEFAULT_USER_AGENT;
use serde_json::{json, Value};
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;

/// A JSON-RPC 2.0 client over HTTP POST. Request ids auto-increment per client.
#[derive(Debug, Clone)]
pub struct JsonRpcClient {
    url: String,
    http: reqwest::Client,
    next_id: Arc<AtomicI64>,
}

impl JsonRpcClient {
    /// Creates a JSON-RPC client targeting the given URL using a fresh HTTP
    /// client.
    pub fn new(url: impl Into<String>) -> Self {
        Self::with_client(url, reqwest::Client::new())
    }

    /// Creates a JSON-RPC client targeting the given URL using the supplied HTTP
    /// client.
    pub fn with_client(url: impl Into<String>, http: reqwest::Client) -> Self {
        Self {
            url: url.into(),
            http,
            next_id: Arc::new(AtomicI64::new(1)),
        }
    }

    /// The configured endpoint URL.
    pub fn url(&self) -> &str {
        &self.url
    }

    /// Invokes a JSON-RPC method and returns its `result`.
    ///
    /// Returns [`Error::JsonRpc`] when the response carries an error member, or
    /// [`Error::Http`] on a non-2xx transport response.
    pub async fn call(&self, method: &str, params: Value) -> Result<Value> {
        let id = self.next_id.fetch_add(1, Ordering::Relaxed);
        let params = match params {
            Value::Null => json!([]),
            other => other,
        };
        let body = json!({
            "jsonrpc": "2.0",
            "id": id,
            "method": method,
            "params": params,
        });

        let resp = self
            .http
            .post(&self.url)
            .header("Content-Type", "application/json")
            .header("Accept", "application/json")
            .header("User-Agent", DEFAULT_USER_AGENT)
            .json(&body)
            .send()
            .await?;

        let status = resp.status();
        let text = resp.text().await?;
        if !status.is_success() {
            return Err(Error::Http {
                status: status.as_u16(),
                url: self.url.clone(),
                body: text,
            });
        }

        let parsed: Value =
            serde_json::from_str(&text).map_err(|e| Error::InvalidResponse(e.to_string()))?;
        if let Some(err) = parsed.get("error").filter(|e| !e.is_null()) {
            let code = err.get("code").and_then(Value::as_i64).unwrap_or(0);
            let message = err
                .get("message")
                .and_then(Value::as_str)
                .unwrap_or("unknown JSON-RPC error")
                .to_string();
            return Err(Error::JsonRpc { code, message });
        }
        Ok(parsed.get("result").cloned().unwrap_or(Value::Null))
    }
}
